### PR TITLE
Updated Binary Search docs to show insert into an already sorted list

### LINKED ISSUE #61684: `binary_search` docs should show how to insert something into an already-sorted list, while maintaining sort order
This isn't exactly hard to do, but it's completely possible to have an off-by-one error because of overthinking it, and it's a common enough use case that it seems worth showing off in a doctest.

### PR DESCRIPTION
**Pull request by ptallo**
Resolves: #61684

### CODE REVIEW

File: src/libcore/slice/mod.rs [hunk 1]
```diff
@@ -1362,6 +1362,23 @@ impl<T> [T] {
     /// let r = s.binary_search(&1);
     /// assert!(match r { Ok(1..=4) => true, _ => false, });
     /// ```
+    /// 
+    /// Insert new element into already sorted array while maintaining sorted
```
Comment (jonas-schievink) [inline src/libcore/slice/mod.rs]:
```suggestion
    /// Inserts a new element into already sorted array while maintaining sorted
```

File: src/libcore/slice/mod.rs [hunk 2]
```diff
@@ -1362,6 +1362,23 @@ impl<T> [T] {
...
+    /// order. Then asserts that the item was inserted correctly
```
Comment (jonas-schievink) [inline src/libcore/slice/mod.rs]:
```suggestion
    /// order. Then asserts that the item was inserted correctly.
```

**Review by jonas-schievink — APPROVED** [review]
Looks good to me, but let's wait for someone on the docs/libs team to look over it. Thanks for the PR :)

**Comment by ptallo** [comment]
No problem! Thanks for the review help, grammar isn't my strong suit!

**Comment by ondj** [comment]
Why something like this?

```rust
if let Err(pos) = s.binary_search(&num) {
    s.insert(pos, num);
}
```

**Comment by ptallo** [comment]
was squashing cleanup commits into one commit, and @ondj I believe that would work as well, this way just seemed like it would allow more modification (editing of the OK() or Err() blocks)

**Comment by czipperz** [comment]
Ping @ptallo

File: src/libcore/slice/mod.rs [hunk 3]
```diff
@@ -1362,6 +1362,23 @@ impl<T> [T] {
...
+    ///
+    /// Inserts a new element into already sorted array while maintaining sorted
+    /// order. Then asserts that the item was inserted correctly.
+    ///
+    /// ```
+    /// let mut s = vec![0, 1, 1, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55];
+    /// let num = 14;
+    ///
+    /// match s.binary_search(&num) {
+    ///     Ok(_) => {}
```
Comment (Mark-Simulacrum) [inline src/libcore/slice/mod.rs]: Hm, I think we should be inserting regardless of finding the item here, so filling out this `Ok` branch with an insert as well would be good, perhaps even illustrating that by doing something like `Ok(idx) | Err(idx) => s.insert(idx, num)` in an if let.

**Comment by Mark-Simulacrum** [comment]
r? @Mark-Simulacrum 

I'll take over review here for now, left a comment.

File: src/libcore/slice/mod.rs (hunk 3 above)
Comment (ptallo) [reply src/libcore/slice/mod.rs]: I believe my new commit is more what you were looking for. Do I need to squash the commits or will the merge do that for me?

Comment (Mark-Simulacrum) [reply src/libcore/slice/mod.rs]: You'll need to squash the commits; it also looks like you have a build failure here -- I suspect the `;` after the if let.

File: src/libcore/slice/mod.rs [hunk 4]
```diff
@@ -1362,6 +1362,21 @@ impl<T> [T] {
...
+    ///
+    /// Inserts a new element into already sorted array while maintaining sorted
+    /// order. Then asserts that the item was inserted correctly.
+    ///
+    /// ```
+    /// let mut s = vec![0, 1, 1, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55];
+    /// let num = 14;
+    ///
+    /// let idx = s.binary_search(&num);
+    /// if let Ok(idx) | Err(idx) {
```
Comment (czipperz) [inline src/libcore/slice/mod.rs]: I believe this is invalid syntax and should be `if let Ok(idx) | Err(idx) = idx {`

Comment (czipperz) [reply src/libcore/slice/mod.rs]: See error:

```
error: expected one of `=` or `|`, found `{`
 --> slice/mod.rs:1361:27
  |
8 | if let Ok(idx) | Err(idx) {
  |                           ^ expected one of `=` or `|` here
```

File: src/libcore/slice/mod.rs [hunk 5]
```diff
@@ -1362,6 +1362,21 @@ impl<T> [T] {
...
+    ///    s.insert(idx, num)
```
Comment (NOVA-ME) [inline src/libcore/slice/mod.rs]: There have been losed a semicolon
s.insert(idx,num);

**Comment by gagan0723** [comment]
ping from triage, @ptallo any updates on this?

**Comment by chocol4te** [comment]
Hello! Second ping from triage, @ptallo please feel free to reopen when you have the time (Also note that pushing to the PR while it is closed prevents it from being reopened).

@rustbot modify labels to +S-inactive-closed, -S-waiting-on-author

**Comment by lzutao** [comment]
I rework in #63442.

**Pull request closed without merging**